pkg/controllers/node/health/controller.go

```
@@ -0,0 +1,129 @@
+			// found unhealthy condition on the node
+			foundCloudProviderPolicy = policy
+			break
+		}
+	}
+
+	// From here there are three scenarios to handle:
+	// 1. If node is healthy, exit node repair loop
+	if foundCloudProviderPolicy.Type == "" {
+		return reconcile.Result{}, nil
+	}
+
+	// 2. If the Node is unhealthy, but has not reached it's full toleration disruption, exit the loop
+	disruptionTime := nodeHealthCondition.LastTransitionTime.Add(foundCloudProviderPolicy.TolerationDuration)
+	if c.clock.Now().Before(disruptionTime) {
+		return reconcile.Result{RequeueAfter: disruptionTime.Sub(c.clock.Now())}, nil
+	}
+
+	if err := c.annotateTerminationGracePeriod(ctx, nodeClaim); err != nil {
+		return reconcile.Result{}, fmt.Errorf("annotated termination grace period on nodeclaim, %w", err)
+	}
+
+	// 3. Otherwise, if the Node is unhealthy and past it's tolerationDisruption window we can forcefully terminate the node
+	if err := c.kubeClient.Delete(ctx, nodeClaim); err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+	// 4. The deletion timestamp has successfully been set for the Node, update relevant metrics.
+	log.FromContext(ctx).V(1).Info("deleting unhealthy node")
+	metrics.NodeClaimsDisruptedTotal.Inc(map[string]string{
+		metrics.ReasonLabel:       metrics.UnhealthyReason,
+		metrics.NodePoolLabel:     nodeClaim.Labels[v1.NodePoolLabelKey],
+		metrics.CapacityTypeLabel: nodeClaim.Labels[v1.CapacityTypeLabelKey],
+	})
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) annotateTerminationGracePeriod(ctx context.Context, nodeClaim *v1.NodeClaim) error {
+	stored := nodeClaim.DeepCopy()
+	terminationTime := c.clock.Now().Format(time.RFC3339)
+	nodeClaim.ObjectMeta.Annotations = lo.Assign(nodeClaim.ObjectMeta.Annotations, map[string]string{v1.NodeClaimTerminationTimestampAnnotationKey: terminationTime})
+
+	if err := c.kubeClient.Patch(ctx, nodeClaim, client.MergeFrom(stored)); err != nil {
+		return client.IgnoreNotFound(err)
```

If we failed to patch due to NotFound, do we really want to ignore the error at this layer? I think we should just bail out of the provisioning loop in general when we get a NodeClaim notFound since there's nothing we can really do at that point